### PR TITLE
Add results, iterables, iterators and indexers, fix options for Kotlin

### PR DESCRIPTION
**Pull request by jcrist1**
This PR adds result types, iterables, iterators and indexers for kotlin as well as fixing option types. It also builds the example with a little test as well. I'm still having trouble getting a high level overview. I think I might be get that overview if I implement an alternate backend (e.g. with Panama).  I also addressed a small change to how `DiplomatWrite` cleanup was handled.

### CODE REVIEW

**Comment by Manishearth** [comment]
> I'm still having trouble getting a high level overview

Of what, exactly?

**Review by Manishearth — COMMENTED** [review]
partial review, reached gen_slice_retrn

**Comment by jcrist1** [comment]
> > I'm still having trouble getting a high level overview
> 
> Of what, exactly?

I only mean of my own backend. It's mostly a vibe, but I feel like in many places one has to implicitly know how a given method was called which makes it harder to reason about locally. It would be nice if the code felt like it could only flow one way at a given point which is not a feeling I have right now. Often I felt like I was adding additional parameters to address issues I saw in the generated code in an ad hoc way.

File: tool/src/kotlin/mod.rs [hunk 1]
```diff
@@ -401,51 +564,196 @@ return string"#
...
+            val_name,
+            return_type_modifier,
         };
         opaque_return
             .render()
             .expect("Failed to render opaque return block")
     }
 
-    fn gen_infallible_return<'d>(
+    fn gen_fallible_return<'d>(
```
Comment (Manishearth) [inline tool/src/kotlin/mod.rs:575]: nit: this generates the return type _conversion_, yes? perhaps we should have that in the name (gen_fallible_return_conversion or something). same for other methods

**Comment by Manishearth** [comment]
done reviewing. would like some attempt put into cleaning up the situation where you need lots of gen methods but if it doesn't work out that's fine too

**Comment by jcrist1** [comment]
@Manishearth I managed to remove the dedicated gen methods for the special methods. I still need a way to get the non-optional type for the generators, but this is now a closure right before the special method handling block inside and it is much more compact.  In general I have renamed all of the `gen_return_*` as `gen_return_*_conversion` as most assume the `returnVal` from the native function call, and create a block that converts the `returnVal` but also provides the return statement.

File: tool/src/kotlin/mod.rs [hunk 2]
```diff
@@ -594,23 +902,80 @@ if (returnVal == null) {{
...
+
+        // this should only be called in the special method generation below
+        let non_option_type_name = |return_type: &ReturnType| match return_type {
+            ReturnType::Infallible(ok) | ReturnType::Nullable(ok) => {
+                self.gen_infallible_return_type_name(ok)
+            }
+            ReturnType::Fallible(_, _) => panic!(
+                "non_option_type_name should only be called for a return type that is optional"
+            ),
+        };
```
Comment (jcrist1) [inline tool/src/kotlin/mod.rs:931]: @Manishearth this is the closure that was in the dedicated methods before. I could still drop the separate branch for the fallible method, as there's no need for the check.

Comment (jcrist1) [reply tool/src/kotlin/mod.rs:931]: e.g.
```rust
        let infallible_type_name_pass_through = |return_type: &ReturnType| match return_type {
            ReturnType::Infallible(ok) | ReturnType::Nullable(ok) | ReturnType::Fallible(ok, _) => {
                self.gen_infallible_return_type_name(ok)
            }
        };

```

**Review by Manishearth — APPROVED** [review]

File: tool/src/kotlin/mod.rs (hunk 2 above)
Comment (Manishearth) [reply tool/src/kotlin/mod.rs:931]: I'd say up to you. I don't mind having duplicate checks as long as they're not overcomplicating the code.

**Comment by Manishearth** [comment]
I'm merging this but you can tweak the check you asked about in another PR if you'd like!

**Pull request merged**